src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs

```diff
@@ -24,26 +19,39 @@ namespace Microsoft.AspNetCore.Server.Kestrel;
 /// </summary>
 public class KestrelConfigurationLoader
 {
+    private ITlsConfigurationLoader? _tlsLoader;
     private bool _loaded;
 
     internal KestrelConfigurationLoader(
         KestrelServerOptions options,
         IConfiguration configuration,
-        IHostEnvironment hostEnvironment,
         bool reloadOnChange,
-        ILogger<KestrelServer> logger,
-        ILogger<HttpsConnectionMiddleware> httpsLogger)
+        ITlsConfigurationLoader? tlsLoader)
     {
-        Options = options ?? throw new ArgumentNullException(nameof(options));
-        Configuration = configuration ?? throw new ArgumentNullException(nameof(configuration));
-        HostEnvironment = hostEnvironment ?? throw new ArgumentNullException(nameof(hostEnvironment));
-        Logger = logger ?? throw new ArgumentNullException(nameof(logger));
-        HttpsLogger = httpsLogger ?? throw new ArgumentNullException(nameof(logger));
-
+        Options = options;
+        Configuration = configuration;
+        ConfigurationReader = new ConfigurationReader(configuration);
         ReloadOnChange = reloadOnChange;
 
-        ConfigurationReader = new ConfigurationReader(configuration);
-        CertificateConfigLoader = new CertificateConfigLoader(hostEnvironment, logger);
+        _tlsLoader = tlsLoader;
+    }
+
+    internal bool IsTlsConfigurationLoadingEnabled => _tlsLoader is not null;
+
+    internal void EnableTlsConfigurationLoading(
+        IHostEnvironment hostEnvironment,
+        ILogger<KestrelServer> serverLogger,
+        ILogger<HttpsConnectionMiddleware> httpsLogger)
+    {
+        if (_tlsLoader is null)
+        {
+            _tlsLoader = new TlsConfigurationLoader(hostEnvironment, serverLogger, httpsLogger);
+
+            if (_loaded)
+            {
+                Reload();
+            }
```

Comment:
Given that if this is called, we know that `UseHttps()` is going to read the default HTTPS config immediately via `ApplyHttpsDefaults` and `ApplyDefaultCertificate`, wouldn't it be better to always reload?

That might be enough to close #45801 with this PR. It's not as good as delaying the configuration reading all the way until binding, but it fixes the primary issue without breaking [dotnet-monitor detecting the lack of default cert by catching InvalidOperationExceptions](https://github.com/dotnet/dotnet-monitor/blob/f3331675bc328332dde4022469fb4b66a44ec83c/src/Tools/dotnet-monitor/AddressListenResults.cs#L92-L113).

```suggestion
            Reload();
```

Reply:
I'm not sure I follow - what would be the benefit of reloading here, rather than where we were about to do it anyway?  When we were discussing #45801, I thought it was important to defer the code-based configuration, rather than to move up the `IConfiguration`-based configuration.

Reply:
> When we were discussing https://github.com/dotnet/aspnetcore/issues/45801, I thought it was important to defer the code-based configuration, rather than to move up the IConfiguration-based configuration.

I still think it's better in principle to delay the code-based configuration to avoid extra config loading. But we tried that, and it broke `dotnet-monitor`.

I'm just trying to be realistic, and in the absence of an easy path forward to delay reading the configuration for `UseHttps()`, recognize that moving up the configuration loading in this case where we know it's going to be eagerly read is still better than not despite knowing it will get reloaded again later.

In a perfect world, we wouldn't be eagerly reading config, but we are. If and when we do delay reading config, I fully support removing the extraneous load.

Reply:
In that case, I think I'd rather leave it the way it is and then make a second change after this is merged removing the `_loaded` check and explaining how that mitigates #45801.  I think that will be much clearer if/when someone tries to clean it up.

Reply:
Are we going to test this by manually calling `kso.ConfigurationLoader!.Load()` before calling `kso.Listen(..., lo => UseHttps())`? Because that's the only way this code ever does anything at all, right?

It seems weird to me to touch this code but not try to make the common case work when it's less code just because it happens to fix another existing bug. That said, if you think this is going to cause controversy or unnecessary friction with this PR, fine. I just don't want to leave this completely unfixed in .NET 8, and putting it off always increases that risk.

Reply:
I wasn't aware that `_loaded` would always be false here.  Can I assert that instead of checking it?

I thought the case being fixed in #45801 was explicitly calling `UseHttps` when running in a docker container.  I assumed that was less common that configuring it by setting the address because I didn't think it had ever worked and it would have been weird to ship with the common case broken.  But maybe you meant something else by "common case"?

Besides the logical tidiness of having separate commits for separate intended outcomes, the reason I'm inclined to separate them is that both the pay-for-play change and the configuration-loading-sequence change seem likely to have unintended consequences, especially in codebases with very specific expectations, and I think investigating and fixing those problems will be easier if the commits are separate.